doc/api18970.md

```diff
@@ -0,0 +1,175 @@
+A thread is pre-empted when it transitions from *Running* to *Ready*
+without having yielded.  (Only preemptible threads can be pre-empted.)
+
+### yielding (function)
+
+A function is yielding when it invokes (directly or indirectly) an
+operation that causes the current thread to yield.
+
+A function is non-yielding when it is not yielding.
+
+### thread-safe
+
+A function is thread-safe if its behavior is correct when invoked from
+multiple threads at the same time.
+
+#### Commentary
+
+Most public API will satisfy this condition; some private API may not.
+
+### reentrant
+
+A function is reentrant if its behavior is correct when it is invoked by
+(indirect) recursion from the same thread.
+
+### interrupt-safe
+
+A function is interrupt-safe if its behavior is not affected by
+concurrent access from interrupts.
+
+#### Commentary
+
+Most public API will satisfy this condition; some private API may not.
+
+We need to be able to say succinctly "Unless otherwise specified all API
+functions are interrupt-safe" and expect people to know what that means.
+A specific example would be the GPIO API.  Because GPIO write functions
+may be invoked from ISRs read-modify-write code like:
+
+```
+u32_t out = gpio->OUT;
+gpio->OUT ^= (out & ~mask) | (value & mask);
+```
+
+*must* be wrapped in a spin-lock to be interrupt-safe.  Many current
+implementations do not satisfy this requirement.
+
+On the other hand internal functions may be written to assume they are
+called with interrupts disabled, or a specific lock held.
+
+**TODO** standard marking?
+
+### atomic
+
+An operation is atomic if the steps it makes internally cannot be
+affected by nor visible to interleaving executions, such as from
+interrupts or thread pre-emption.
+
+An operation that is atomic is by definition interrupt-safe.
+
+An operation that is atomic is by definition thread-safe.
+
+### asynchronous
+
+A function is asynchronous if it may return before the operation it
+initiates has completed.  An asynchronous function must provide a
+mechanism by which completion is reported, e.g. a callback, event, or
+signal.
+
+#### Commentary
+
+Note that asynchronous is orthogonal to non-yielding.  Some API may
+provide completion information through a callback, but may yield while
+waiting.
+
+### queued
+
+A function is queued if it is asynchronous and allows multiple
+operations to be outstanding at any time.
+
+#### Commentary
+
+This concept is necessary because of operations like
+`spi_transceive_async()` which returns its result through a signal, but
+will yield if the device is already processing an asynchronous
+operation.
+
+Queued operations are rare(? do not exist?) in Zephyr, and require that
+the API use a chainable persisted state object to hold the operation
+parameters in a persisted state object that can be added to an internal
+queue for processing when the required resource is available.
+
+## Other Rules
+
+* A function that cannot fulfill its contract when invoked from
```

Comment:
Are you sure the overhead is worth it for this one? I'm not convinced it is -- if we document that it's not OK to call something from ISR context, the results of doing so should be UB, I think.

Reply:
Obviously this needs to be clarified.  I mean that if the function is documented to be callable from an ISR (as, for example, `gpio_pin_write()` should be), then when that generic interface is invoked from an ISR on a particular implementation (an I2C-based GPIO peripheral) that doesn't allow it, the behavior of the API must not only be defined, but be exactly this behavior.  I.e. it is callable, but it may not succeed.

A function that is documented to not be OK to call from an ISR can produce undefined behavior if that restriction is ignored.

Reply:
This has been addressed by introduction of and commentary on *no-wait*.